Suppress -Wsuggest-override warnings in user code for GCC the same as…

obfuscated opened the pull request:
… as is done for -Winconsistent-override for clang

* This should be done also for clang when it gets the same warning in an
  actual release.

obfuscated committed:
Suppress -Wsuggest-override warnings in user code for GCC the same as as is done for -Winconsistent-override for clang

* This should be done also for clang when it gets the same warning in an
  actual release.

obfuscated commented:
I guess it would be useful if the -Werror=suggest-override flag is added at build time, but currently it fails to build external libraries like scintilla.

If there is an easy way to exclude whole files and/or folders I would be happy to do some testing.

obfuscated commented:
Hm, annoying.

There are two problems:
1. Old GCC compiler - this is easy. Should I make a macro like wxGCC_WARNING_SUPPRESS_CAST_FUNCTION_TYPE for this warning, too?
2. It seems this warning affects clang, but it shouldn't IMO, this is a GCC specific suppression after all. Is there a reason for GCC suppressions to be applied when building with clang? What would I break if I remove this?

vadz commented:
Most of gcc warnings exist(ed, when these macros were added) in clang too, so it's usually convenient to be able to use a single macro to cover both of them. It looks like we may need some `wxONLY_GCC_WARNING_XXX` too now...

In any case, I think it would be useful to have `wxWARNING_SUPPRESS_MISSING_OVERRIDE` which would do the right thing for both compilers (and nothing for older gcc versions). Could you please update the PR to use such a macro (with a better name if you have any suggestions)? TIA!

obfuscated committed:
Fix build with clang compilers

* wxGCC_WARNING_SUPPRESS works also for clang, so we're trying to suppress a
  warning which is not present in most versions of clang.
* With this commit I'm adding a special macro to suppress this warning. This
  macro would work with future versions of clang (probably 11) which would
  support this warning.

obfuscated commented:
I've tried to implement the suggested macro for both clang and gcc. Lets see if I've succeeded.

vadz commented on include/wx/defs.h:
```diff
@@ -708,6 +708,18 @@ typedef short int WXTYPE;
 
+/* Specific maros for -Wsuggest-override warning, new in gcc 5. */
+#if defined(__clang__)
+#   define wxWARNING_SUPPRESS_MISSING_OVERRIDE() wxCLANG_WARNING_SUPPRESS(suggest-override)
```
Wait, shouldn't this be `inconsistent-missing-override`?

vadz commented on include/wx/event.h:
```diff
@@ -4274,8 +4274,10 @@ typedef void (wxEvtHandler::*wxPressAndTapEventFunction)(wxPressAndTapEvent&);
         static const wxEventTableEntry sm_eventTableEntries[];          \
     protected:                                                          \
         wxCLANG_WARNING_SUPPRESS(inconsistent-missing-override)         \
```
The idea was to make this unnecessary.

obfuscated replied:
Clang 11 will get suggess-override warning, too (in version 11) and it will have two separate warnings for this.
wxCLANG_WARNING_SUPPRESS uses __has_warning to check if a warning is available, so it works.

I suppose, I can add the inconsistent-missing-override, but this wasn't my goal.

vadz replied:
Could you please still add it, maybe only for clang < 11 if the new warning will replace the old one in the latter? It's definitely not ideal to have to use 2 different macros for disabling essentially the same warning, just called differently. TIA!

obfuscated replied:
OK. Done in b14d085

vadz reviewed (commented):
Sorry, it seems that this is still not correct for clang.

obfuscated committed:
Make wxWARNING_SUPPRESS_MISSING_OVERRIDE to suppress -Winconsistent-missing-override, too

obfuscated commented:
Do I need to do anything else for this PR?

vadz commented:
No, thanks, this looks good to me now and I'll merge it in a moment.
